### DISCUSSION TITLE
relationship(lazy="raise") with cascaded deletes and field assignments

### DISCUSSION THREAD

**Discussion by mtvx** [sqlalchemy/sqlalchemy#9528] [Usage Questions]
When using `relationship(..., lazy="raise")` following cases become somewhat problematic:

```py
class Parent(Base):
    ...
    children = relationship(Child, back_populates="parent", cascade="all, delete-orphan", lazy="raise")

class Child(Base):
    ...
    parent = relationship(Parent, back_populates="children")

with db.session() as session:
    parent = Parent(...)
    parent.children = [child1]  # raises

    ...

    session.delete(parent)  # raises
```

So, the problem is that when you attempt to set the backref, or delete the parent which configures cascade, error is risen. That's understandable, that's why "raise" is there for, but is there any way to maybe disable it afterwards? Can be done when querying, but how about above cases?

Explicitly loading up all the backrefs may be too tedious work, for deep hierarchies, and sometimes (e.g. tests, scripts, simple cases) lazy loading for deletion is just fine, even if you by general may want to turn off lazy loading via `relationship(lazy="raise")`.

**Comment by zzzeek** [comment]
>  but is there any way to maybe disable it afterwards?

I'm not following here what you want to "disable afterwards",  do you mean some way to disable the "raise" for the duration that you are setting the collection?   I think loading the collection up front would be the best way so that it doesn't have to go out and fetch.    The ORM has no other way to know, when assigning a new collection, what rows need to be de-associated, unless they are loaded.  if you can illustrate more explicitly what you mean that would be helpful.

For the session.delete() case, you should be able to use passive_deletes for that one since foreign key ON DELETE CASCADE can handle that, e.g. 

      children = relationship("Child", back_populates="parent", cascade="all, delete-orphan", lazy="raise", passive_deletes=True)

see https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-foreign-key-on-delete-cascade-with-orm-relationships for that use case.

**Reply by mtvx** [reply]
Thank you for the reply and for maintaining such a great product over the years.

1. For the cascade case, I _don't_ have `ON DELETE CASCADE` in the DB. I still need SQLAlchemy to lazy-load and drop the relationships in this case.

   Consider a deeply nested hierarchy. Writing a query that `.options(selectinload(...))`s everything would become lengthy and you'd need to update it each time a new model is added to the hierarchy. For this reason, at least in tests I would like to drop an entity by a simple `session.delete()`. But the `relationship(lazy="raise")` blocks that, so I'm wondering if I could "disable" the `lazy="raise"` temporarily, just for the delete.

   This may sound like a corner case... but it would be convenient sometimes (e.g. tests where being so "proper" all the time isn't needed).

2. For the assignment:

   > do you mean some way to disable the "raise" for the duration that you are setting the collection?

   Yes, exactly. Thank you for pointing out that SQLAlchemy needs to clear the existing children in the DB. Makes sense that the error is risen.

   However, when you're _creating_ entries, e.g. `parent = Parent(); parent.children = [child]` the model is not yet persisted in the DB, and there's nothing to load and drop. Maybe it wouldn't need to raise here?

   Again, maybe a little bit of a corner case.

**Reply by zzzeek** [reply]
> Thank you for the reply and for maintaining such a great product over the years.
> 
>     1. For the cascade case, I _don't_ have `ON DELETE CASCADE` in the DB. I still need SQLAlchemy to lazy-load and drop the relationships in this case.
>        Consider a deeply nested hierarchy. Writing a query that `.options(selectinload(...))`s everything would become lengthy and you'd need to update it each time a new model is added to the hierarchy. For this reason, at least in tests I would like to drop an entity by a simple `session.delete()`. But the `relationship(lazy="raise")` blocks that, so I'm wondering if I could "disable" the `lazy="raise"` temporarily, just for the delete.
>        This may sound like a corner case... but it would be convenient sometimes (e.g. tests where being so "proper" all the time isn't needed).

well the problem with some kind of "disable all raiseload" parameter becomes that we dilute the "raiseload" case, the operation might now cause hundreds of lazy loads and the programmer might still be surprised by that, preferring some way to assert the relationships that are needed for the delete to proceed.  This is also a "corner case", but we're in "corner case" territory already :) .   

When you say "writing a query that selectinloads everything would require explicit updating for model updates" that's not really true because you can use wildcards for loader options and you can also use get(), like this:

```py
with Session(e) as sess:
    p1 = sess.scalars(select(Parent)).one()

    sess.get(Parent, p1.id, populate_existing=True, options=[selectinload('*')])

```

Full disclosure I had to run a test script and dig into our test suite to find coverage to make sure that the wildcard works recursively, since the [docs](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#wildcard-loading-strategies) don't seem to make that clear.

If we compare that admittedly more verbose query above to the case of `session.delete(obj, disable_all_raiseload=True)`, we can see what we get for this price which is that for a deeply nested hierarchy we get much more efficient SQL queries by indicating a full load up front, rather than having `session.delete()` crawl the hierarchy after the fact emitting loads as needed.

Re: the verbosity of get(), we have a `session.refresh()` method that can also load relationships that are normally lazy='raise', however it doesnt accept loader options and wouldn't handle arbitrary recursiveness.   There is now a lot of functional overlap between `session.get()` and `session.refresh()`, and you can in fact use `get()` for everything that `refresh()` does, so improving the API of `get()` in some way to provide a shorthand for the above, like `session.get_existing(object, options=...)` might be helpful, without breaking the contract of "you can tell us how to load things" provided by the options interface.  This could someday lead to `Session.refresh()` being set up as legacy status, since we dont like having two methods that do the same thing in different ways.


> 
>     2. For the assignment:
>        > do you mean some way to disable the "raise" for the duration that you are setting the collection?
>        
>        
>        Yes, exactly. Thank you for pointing out that SQLAlchemy needs to clear the existing children in the DB. Makes sense that the error is risen.
>        However, when you're _creating_ entries, e.g. `parent = Parent(); parent.children = [child]` the model is not yet persisted in the DB, and there's nothing to load and drop. Maybe it wouldn't need to raise here?

it does not raise if you just created `Parent()`.  If `Parent` is not in the [persistent](https://docs.sqlalchemy.org/en/20/glossary.html#term-persistent) state then no loading occurs.

However, a common pitfall here is that the programmer will create "Parent()" and then persist the object, *then* add the .children, that will raise, but again, only if the .children collection is not already present:

```py
with Session(e) as sess:
    # p1 is transient
    p1 = Parent()

    # now p1 is pending
    sess.add(p1)

    # flush p1, now it's persistent
    sess.flush()

    # now this would raise
    p1.children = [Child(sub_children=[SubChild()]), Child()]

```

`.children` being unloaded above still warrants a lazy load, because the `Session` does not know that you didn't add `Child()` objects linking to this parent in some way separately.   because the collection wasn't stated, it does not assume it's empty.

The solution to this, which is very much assisted by the new [dataclasses](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping) feature that allows for default collections to be generated in constructors, is to assign the empty .children collection up front so that nothing needs to be loaded:

```py
with Session(e) as sess:
    # p1 is transient
    p1 = Parent(children=[])

    # now p1 is pending
    sess.add(p1)

    # flush p1, now it's persistent
    sess.flush()

    # .children collection is already present, is replaced without any lazy loading
    p1.children = [Child(sub_children=[SubChild()]), Child()]

    sess.commit()
```

you can also add an `__init__()` method to `Parent` that sets this blank collection automatically if not present:

```py
class Parent(Base):
    def __init__(self, **kw):
        kw.setdefault('children', [])
        super().__init__(**kw)

```


for reference here's the test case I'm working with, since I test all these cases before putting them out (I'm always surprised by things I've forgotten or have changed, so always need to test)

```py
from sqlalchemy import *
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import *

Base = declarative_base()


class Parent(Base):
    __tablename__ = "parent"

    def __init__(self, **kw):
        kw.setdefault('children', [])
        super().__init__(**kw)

    id = Column(Integer, primary_key=True)

    children = relationship(
        "Child",
        cascade="all, delete-orphan",
        lazy="raise",
    )


class Child(Base):
    __tablename__ = "child"

    id = Column(Integer, primary_key=True)
    parent_id = Column(ForeignKey("parent.id"))
    sub_children = relationship(
        "SubChild", lazy="raise", cascade="all, delete-orphan"
    )


class SubChild(Base):
    __tablename__ = "subchild"

    id = Column(Integer, primary_key=True)
    parent_id = Column(ForeignKey("child.id"))


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

# fixture data
with Session(e) as sess:
    sess.add(Parent(children=[Child(sub_children=[SubChild()]), Child()]))
    sess.commit()

# test one - recursively eagerload so that we can delete
with Session(e) as sess:
    p1 = sess.scalars(select(Parent)).one()

    sess.get(
        Parent, p1.id, populate_existing=True, options=[selectinload('*')]
    )

    # one-level-deep support is there for refresh()
    #sess.refresh(p1, attribute_names=inspect(p1).mapper.relationships.keys())

    p1.children = []

    sess.commit()

# test two - a pending Parent wont lazyload
with Session(e) as sess:
    # p1 is transient
    p1 = Parent()

    # now p1 is pending
    sess.add(p1)

    # for transient or pending, no issue setting new collection
    p1.children = [Child(sub_children=[SubChild()]), Child()]

    sess.commit()

# test three - the common "gotcha" for this case
with Session(e) as sess:
    # p1 is transient
    p1 = Parent()

    # now p1 is pending
    sess.add(p1)

    # flush p1, now it's persistent
    sess.flush()

    # now this would raise if .children wasn't pre-assigned
    p1.children = [Child(sub_children=[SubChild()]), Child()]

    sess.commit()

```

**Reply by mtvx** [reply]
Hey, thanks for the detailed response!

1. Deletions

   - Wouldn't the wildcard loading strategy, e.g. `selectinload('*')`, load up the _whole hierarchy_ and not only those which have deletion cascades configured? This could lead to loading lots of unnecessary data. Could we `selectinload('*', for_deletes=True)` so that it overrides only when delete cascade is configured...?

     Oh, please do add that "recursively" to the docs 😄 I was wondering about the same reading that up!

   - Ability to "eagerly delete" feels potentially important. It should benefit both lazy-loaders and lazy-disablers, making deletes (sometimes considerably) faster.

     I would normally choose to use DB level cascades, but one can't always decide (legacy project, reasons X and Y, etc).

2. Assignments

   - Thanks for correcting me on when the `parent.children` assignment raises. Whoopsie.

   - Constructor/constructing tips acked!

(Pasting the whole "test case" in the response is transparent and nice 👍🏻)

**Reply by zzzeek** [reply]
> Hey, thanks for the detailed response!
> 
>     1. Deletions
> 
> 
>     * Wouldn't the wildcard loading strategy, e.g. `selectinload('*')`, load up the _whole hierarchy_ and not only those which have deletion cascades configured? This could lead to loading lots of unnecessary data. Could we `selectinload('*', for_deletes=True)` so that it overrides only when delete cascade is configured...?

here's what should work, but doesn't, which is to use `lazyload('*')`.   The recursiveness of  `'*'` for relationship loaders is inconsistent.  #9548 is added where we will add a new recursive wildcard `'**'` for this purpose; `lazyload('**')`, `selectinload('**')`.   The case of `selectinload('*')` will emit a warning when traversing more than one level deep that the recursiveness here is deprecated; to remove the warning, use `selectinload('**')` for truly recursive or `selectinload('*|')` (or something) for non-recursive.  in 2.1, `'*'` will act like `'*|'`, something like that.

> 
>     * Ability to "eagerly delete" feels potentially important. It should benefit both lazy-loaders and lazy-disablers, making deletes (sometimes considerably) faster.

delete() should be disabling lazy='raise' in any case.  Right now it's not consistent with the changes made in #5984, dont do lazy='raise' inside of a flush, and particularly #8862, where we disable lazy='raise' inside of `merge()`.  if `merge()` disables lazy='raise' then `delete()` certainly should as well.  #9549 is added

here's the kind of API we don't want:

    do_a().do_b().do_c(but_do_it_this_way=True).do_d()...

that is, the keyword arg for operations that should be very simple verbs is the antipattern.

**Reply by mtvx** [reply]
- Right, `lazyload('*')` workaround should do it, I'll give it a try
- `**` vs `*`, will be great to distinguish between the two 👍🏻 (`*|` looks quite odd, but I guess that's just to retain backwards-compatibility with the `*`)
- https://github.com/sqlalchemy/sqlalchemy/issues/9549 🥳 

I believe this concludes the discussion here. I'll follow https://github.com/sqlalchemy/sqlalchemy/issues/9549. Thanks again!
